### DISCUSSION TITLE
TikZJax not rendering

### DISCUSSION THREAD

**Discussion by AwesomeQuest** [fonsp/Pluto.jl#3069] [Questions about using Pluto] [answered]
I'm trying to use TikZJax inside a notebook but nothing is rendering.
I'm putting this into a cell in the notebook
```
html"""
<p> head </p>
<head>
	<link rel="stylesheet" type="text/css" href="https://tikzjax.com/v1/fonts.css">
	<script src="https://tikzjax.com/v1/tikzjax.js"></script>
</head>
<p> head </p>
<body>
	<script type="text/tikz">
		\begin{tikzcd}
			A \arrow[r, "\phi"] \arrow[d, red]
			  & B \arrow[d, "\psi" red] \\
			C \arrow[r, red, "\eta" blue]
			  & |[blue, rotate=-15]| D
		\end{tikzcd}
	</script>
<p> head </p>
</body>
"""
```
And when this code is put into a html file it looks like this
![image](https://github.com/user-attachments/assets/0f7cf75e-c153-4991-a40c-99d4430fe387)

But inside the notebook I only get this
![image](https://github.com/user-attachments/assets/eeae6af6-bdd4-4a0d-8f53-f8cca797b4cc)

What's going wrong?

What TikZJax is supposed to do is take the bit enclosed inside <script type="text/tikz"> tag and compile it to svg and then display that svg. But when inspecting the Pluto cell the script tag remains unchanged

Does Pluto.jl not allow the output of it's cells to be modified by the javascript in other cells? Since I think the code that does the compiling is inside the header so is that the problem?

I don't think it's TikZJax since when I copy the html of the pluto-cell tag into another html file it renders fine, meaning I don't think wrapping the <script type="text/tikz"> in other tags is not the problem.

**Comment by fonsp** [comment]
Pluto allows js to modify all content, but i suspect that tikzjax is only designed to convert scripts into diagrams once, during page load, and not when added dynamically like you are doing.

Check out the featured notebooks about JavaScript to learn more about how to debug this.

You can also tell Pluto to render your cell in an iframe: put an <html> tag around the content.

**Reply by AwesomeQuest** [reply, marked as answer]
Thanks! That seems to work! I threw together a macro as a proof of concept.
```
macro tikz_str(input, height)
	base = """
	<head>
		<link rel="stylesheet" type="text/css" href="https://tikzjax.com/v1/fonts.css">
		<script src="https://tikzjax.com/v1/tikzjax.js"></script>
	</head>
	<body>
	<script type="text/tikz">
		$input
	</script>
	</body>
	"""
	return @htl("""
          <iframe srcdoc='
          $base
          ' width=100% height=$height></iframe>
          """)
end

tikz"""
\huge
\begin{tikzcd}
	A \arrow[r, "\phi"] \arrow[d, red]
	  & B \arrow[d, "\psi" red] \\
	C \arrow[r, red, "\eta" blue]
	  & |[blue, rotate=-15]| D
\end{tikzcd}
"""200
```

![image](https://github.com/user-attachments/assets/5a3ee9d5-b0d3-4b10-bb08-d374f85488bb)

**Reply by fonsp** [reply]
That looks great!! Thanks for sharing

**Comment by AwesomeQuest** [comment]
Ok, so after some more research and tinkering I've come upon a much better way to integrate Tikz into Pluto.jl. Though it does require a working lualatex in your path to run.

TikzPictures.jl already exists and integrates with IJulia but doesn't quite work out of the box for Pluto.jl and also has almost no documentation so here's my solution.

Thankfully Pluto.jl does simply display the output of TikzPictures.jl but simple Tikz doesn't really work in dark mode and TikzPictures.jl has a rather cumbersome syntax for adding extra packages, using different environments and options. My solution goes as follows:
```julia
begin
  using TikzPictures
  macro tikz_str(input, width=20)
    preamble, environment, options, input = split(input, ';', limit=4)
    environment == "" && (environment = "tikzpicture")
		
    TikzPicture(input, 
        preamble="\\usetikzlibrary{backgrounds} $preamble", 
	options="$options , background rectangle/.style={fill=white!45}, show background rectangle", 
	width="$(width)cm")
  end
end
```

Which allows one to insert tikz into Pluto.jl with a little extra syntax at the top
```julia
tikz"""
\usepackage{circuitikz}
;
circuitikz
;
american voltages
;
\draw
  (0,0) to [short, *-] (6,0)
  to [V, l_=$\mathrm{j}{\omega}_m \underline{\psi}^s_R$] (6,2) 
  to [R, l_=$R_R$] (6,4) 
  to [short, i_=$\underline{i}^s_R$] (5,4) 
  (0,0) to [open, v^>=$\underline{u}^s_s$] (0,4) 
  to [short, *- ,i=$\underline{i}^s_s$] (1,4) 
  to [R, l=$R_s$] (3,4)
  to [L, l=$L_{\sigma}$] (5,4) 
  to [short, i_=$\underline{i}^s_M$] (5,3) 
  to [L, l_=$L_M$] (5,0);
"""
```

![image](https://github.com/user-attachments/assets/ea8b4463-4142-40b4-8cf4-ac57873404b8)


Something like this, in my opinion, completes Pluto.jl's scientific feature set. Making a good diagram is often absolutely necessary. I understand not having support for arbitrary LaTeX, but support for some of the most common tikz packages would be wonderful.

**Reply by fonsp** [reply]
Nice! You say that this is a better solution, maybe you can say a bit how it compares to tikzjax? Also consider making a PR to TikzPictures (documentation or code) or registering your own package (PlutoTikz.jl)! Small packages are very useful!
